doc/source/development/roadmap.rst

```diff
@@ -53,6 +53,32 @@ need to implement certain operations expected by pandas users (for example
 the algorithm used in, ``Series.str.upper``). That work may be done outside of
 pandas.
 
+Consistent missing value handling
+---------------------------------
+
+Currently, pandas has varying missing data interfaces depending on the data
+type: pandas uses ``np.nan`` as missing value indicator in floating point data,
+``np.nan`` or ``None`` in object dtype data (eg strings, or booleans with
+missing values are cast to object), ``pd.NaT`` in datetimelike data. For
+categorical or interval data, they return ``np.nan`` on access even when the
+categories or intervals are datetime-like. Integer data cannot store missing
+data or are cast to float.
+
+Long term, we want to introduce consistent missing value handling accross the
+different data types: all data types should support missing values and with the
```

Comment:
Not sure I agree that all data types should support missing values. Non-nullable types could be beneficial

Reply:
I agree that there's value in pandas ensuring that a column cannot contain NAs. I'm not sure where best to put that invariant: the dtype or the array / column.

But, to sidestep this issue, perhaps something like "pandas should provide consistent missing value handling for all the different data types", i.e. we give you the *ability*, without saying that every *dtype* has to be nullable.

Reply:
I seem to remember that you, Tom and I had a similar discussion in another issues (but can't directly find it). 

I agree that the concept of non-nullability is useful/interesting, but as Tom also mentions:  non-nullability shouldn't necessarily be a property of the data type itself. Similarly as Tom is working on the "allow_duplicate_labels" flag, we could have "nullable=True/False" flags per column. 

Because for the dtype itself to be non-nullable, we should ask ourselves: do we have an example of a data type (that we want to include in pandas) for which it would never be useful to be nullable? 

Or to phrase the text in a different way: *if* a data type supports missing values, it should follow consistent semantics.

---

Note that right now, pandas doesn't really know the concept of non-nullable. Yes, we have some dtypes that don't support NAs (like integer dtype), but whenever some operation introduces a missing value, we simply upcast to a dtype that can store the missing value (so in practice means changing to float for integer). A proper concept of `nullable=False` flag wouldn't necessarily work like this, I think.